src/lib/environment.ts

```diff
@@ -48,40 +49,75 @@ export class Environment<StateT, ActionT extends Action> {
     return selector(this.store.getState(), ...args);
   };
 
-  public run<T, P extends any[]>(f: (...params: P) => T, ...params: P): T;
-  public run<T, P extends any[]>(effect: Effect<StateT, ActionT, P, T>, ...params: P): T;
-  public run<T, P extends any[]>(funcOrEffect: Effect<StateT, ActionT, P, T> | ((...params: P) => T), ...params: P): T {
+  /**
+   * Calls `f` with the supplied params
+   *
+   * In case the return type is not correctly inferred, wrap the function in an BoundEffect
+   *
+   * @param f
+   * @param params
+   */
+  public call = <T, P extends any[]>(f: (...params: P) => T, ...params: P): T => {
     if (this.cancellationToken && this.cancellationToken.canceled) {
       throw new SagaCancelledError(`Saga has been cancelled`);
     }
 
-    if (typeof funcOrEffect === 'function') {
-      return funcOrEffect(...params);
-    } else {
-      return funcOrEffect.func(this, ...params);
-    }
-
-    // return f(...params);
-  }
+    return f(...params);
+  };
 
   public take = async <Payload>(actionCreator: ActionCreator<Payload>): Promise<Payload> => {
     return (await this.waitForMessage(actionCreator)).payload;
   };
 
-  // public spawn = <P extends any[], T>(f: (env: Environment<StateT, ActionT>, ...args: P) => T, ...args: P): T => {
-  //   return f(this, ...args);
-  // };
+  /**
+   * Runs the given saga as an attached child.
+   * Cancelling the parent will also cancel the child at the next opportunity.
+   *
+   * @param effectOrEffectCreator
+   * @param params
+   */
+  public run<T, P extends any[]>(effect: BoundEffect<Environment<StateT, ActionT>, P, T>): T;
+  public run<T, P extends any[]>(effect: SagaFunc<StateT, ActionT, P, T>, ...params: P): T;
+  public run<P extends any[], T>(
```

Comment:
So the only difference between `run` and `call` is now the first arg being the environment?

I think we should rename one of them because the difference is not clear from the name.

Reply:
Yes, that's the difference. Do you have any naming suggestions for these? I want to raise them with the team today.

The main benefit I still see of not further overloading `call` (or whatever name we'll end up with) is that one can gauge the complexity / impact of the call right from the signature, without having to inspect the type whether it receives `Env` or not.